### PR TITLE
Let some processes of cic run as system user

### PR DESCRIPTION
**Pull request by ceiba1985**
For security reason, our customer want all the processes of
a container to run as non-root user, so let some processes
that run as root user previously run as system user.

Tracked-On: OAM-90885
Signed-off-by: ji, zhenlong z <zhenlong.z.ji@intel.com>

### CODE REVIEW

File: android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+         if (prctl(PR_SET_SECUREBITS, securebits) != 0) {
+             PLOG(FATAL) << "prctl(PR_SET_SECUREBITS) failed for " << name_;
+         }
+@@ -916,6 +919,11 @@ Result<Success> Service::Start() {
+             ZapStdio();
+         }
+ 
++        if (strcmp("adbd", name_.c_str()) == 0) {
++            if (getuid() != 0)
++                setuid(0);
```
Comment (joez) [inline android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch:179]: which means adbd will still run as root?

Comment (ceiba1985) [reply android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch:179]: adbd will invoke setuid itself. consider adb root/unroot.

File: android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+diff --git a/init/builtins.cpp b/init/builtins.cpp
+index 8bd92ccdd..147069f29 100644
+--- a/init/builtins.cpp
++++ b/init/builtins.cpp
+@@ -128,9 +128,21 @@ static Result<Success> do_class_restart(const BuiltinArguments& args) {
+ }
+ 
+ static Result<Success> do_domainname(const BuiltinArguments& args) {
++    int ori_uid = getuid();
```
[the remark this answers is not in the document]
Comment (ceiba1985) [reply android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch:34]: Can do this, it just modify an API may introduce more modifications.

File: android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+     android::base::unique_fd fd(TEMP_FAILURE_RETRY(
+         OpenFile(path, O_WRONLY | O_CREAT | O_NOFOLLOW | O_TRUNC | O_CLOEXEC, 0600)));
+     if (fd == -1) {
++        if (ori_uid > 0 && as_root)
++            setuid(ori_uid);
+         return ErrnoError() << "open() failed";
+     }
+     if (!android::base::WriteStringToFd(content, fd)) {
++        if (ori_uid > 0 && as_root)
++            setuid(ori_uid);
```
Comment (joez) [inline android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch:194]:
```cpp
SETUID_END;
```

File: android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+         return ErrnoError() << "Unable to write file contents";
+     }
++    if (ori_uid > 0 && as_root)
++        setuid(ori_uid);
```
Comment (joez) [inline android/common/system/core/0005-Let-some-core-services-run-as-system-user.patch:198]:
```cpp
SETUID_END;
#undef SETUID_START
#undef SETUID_END
```

**Comment by joez** [comment]
a better way to achieve this is create a wrapper function, e.g.

```cpp
static bool SetHighestAvailableOptionValueAsRoot(std::string path, int min, int max) {
  int ori_uid = 0;
  if (as_root && (ori_uid = getuid()) != 0)
      setuid(0);
  bool ret = SetHighestAvailableOptionValue(path, min, max);
  if (as_root && ori_uid != 0)
      setuid(ori_uid);
  return ret;
}
```

**Comment by joez** [comment]
LGTM

**Pull request merged**